x/mongo/driver/auth/aws_conv.go

```diff
@@ -0,0 +1,353 @@
+	err = json.Unmarshal(creds, &es2Resp)
+	if err != nil {
+		return nil, err
+	}
+	ac.username = es2Resp.AccessKeyID
+	ac.password = es2Resp.SecretAccessKey
+	ac.token = es2Resp.Token
+
+	return ac.validateAndMakeCredentials()
+}
+
+func (ac *awsConversation) getCredentials() (*credentials.Credentials, error) {
+	// Credentials passed through URI
+	creds, err := ac.validateAndMakeCredentials()
+	if creds != nil || err != nil {
+		return creds, err
+	}
+
+	// Credentials from environment variables
+	ac.username = os.Getenv("AWS_ACCESS_KEY_ID")
+	ac.password = os.Getenv("AWS_SECRET_ACCESS_KEY")
+	ac.token = os.Getenv("AWS_SESSION_TOKEN")
+
+	creds, err = ac.validateAndMakeCredentials()
+	if creds != nil || err != nil {
+		return creds, err
+	}
+
+	// Credentials from ECS metadata
+	relativeEcsURI := os.Getenv("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+	if len(relativeEcsURI) > 0 {
+		fullURI := awsRelativeURI + relativeEcsURI
+
+		req, err := http.NewRequest("GET", fullURI, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = resp.Body.Close() }()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		var espResp ecsResponse
+		err = json.Unmarshal(body, &espResp)
+		if err != nil {
+			return nil, err
+		}
+		ac.username = espResp.AccessKeyID
+		ac.password = espResp.SecretAccessKey
+		ac.token = espResp.Token
+
+		creds, err = ac.validateAndMakeCredentials()
+		if creds != nil || err != nil {
+			return creds, err
+		}
+	}
+
+	// Credentials from EC2 metadata
+	creds, err = ac.getEC2Credentials()
+	if creds == nil && err == nil {
+		return nil, errors.New("unable to get credentials")
+	}
+	return creds, err
+}
+
+func (ac *awsConversation) firstMsg() ([]byte, error) {
+	// Values are cached for use in final message parameters
+	ac.nonce.Data = make([]byte, 32)
+	_, _ = rand.Read(ac.nonce.Data)
+
+	msg := bsoncore.BuildDocumentFromElements(nil,
+		bsoncore.AppendInt32Element(nil, "p", 110),
+		bsoncore.AppendBinaryElement(nil, "r", 0x00, ac.nonce.Data),
+	)
+	return msg, nil
+}
+
+func (ac *awsConversation) finalMsg(s1 []byte) ([]byte, error) {
+	var sm serverMessage
+	err := bson.Unmarshal(s1, &sm)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check nonce prefix
+	if sm.Nonce.Subtype != 0x00 {
+		return nil, errors.New("server reply contained unexpected binary subtype")
+	}
+	if len(sm.Nonce.Data) != 64 {
+		return nil, errors.New("server reply nonce was not 64 bytes")
+	}
+	if !bytes.HasPrefix(sm.Nonce.Data, ac.nonce.Data) {
+		return nil, errors.New("server nonce did not extend client nonce")
+	}
+
+	region, err := getRegion(sm.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	creds, err := ac.getCredentials()
+	if err != nil {
+		return nil, err
+	}
+
+	tme := time.Now()
```

Comment:
`time.Now` gives the time in the local time zone, so we have to do `time.Now().UTC()` instead.